test_runner/src/main/kotlin/ftl/run/platform/common/AfterRunTests.kt

```diff
@@ -41,3 +48,16 @@ private fun saveConfigFile(matrixMap: MatrixMap, args: IArgs) {
     configFilePath.parent.toFile().mkdirs()
     Files.write(configFilePath, args.data.toByteArray())
 }
+
+private suspend inline fun MatrixMap.printMatricesWebLinks(project: String) = coroutineScope {
+    println("Matrices webLink")
+    map.values.map { launch { it.printWebLink(project) } }.joinAll()
+    println()
+}
+
+private suspend inline fun SavedMatrix.printWebLink(project: String) =
+    println("${FtlConstants.indent}$matrixId: ${getOrUpdateWebLink(webLink, project, matrixId)}")
+
+private tailrec suspend fun getOrUpdateWebLink(link: String, project: String, matrixId: String): String =
+    if (link.isNotBlank()) link
+    else getOrUpdateWebLink(GcTestMatrix.refresh(matrixId, project).webLink(), project, matrixId)
```

Comment:
A matrix is not guaranteed to have a webLink. A matrix can be invalid.
https://github.com/Flank/flank/blob/master/test_runner/src/main/kotlin/ftl/util/MatrixState.kt

Here are the possible states:

```
        "enumDescriptions": [
            "Do not use.  For proto versioning only.",
            "The execution or matrix is being validated.",
            "The execution or matrix is waiting for resources to become available.",
            "The execution is currently being processed.\n\nCan only be set on an execution.",
            "The execution or matrix has terminated normally.\n\nOn a matrix this means that the matrix level processing completed normally,\nbut individual executions may be in an ERROR state.",
            "The execution or matrix has stopped because it encountered an\ninfrastructure failure.",
            "The execution was not run because it corresponds to a unsupported\nenvironment.\n\nCan only be set on an execution.",
            "The execution was not run because the provided inputs are incompatible with\nthe requested environment.\n\nExample: requested AndroidVersion is lower than APK's minSdkVersion\n\nCan only be set on an execution.",
            "The execution was not run because the provided inputs are incompatible with\nthe requested architecture.\n\nExample: requested device does not support running the native code in\nthe supplied APK\n\nCan only be set on an execution.",
            "The user cancelled the execution.\n\nCan only be set on an execution.",
            "The execution or matrix was not run because the provided inputs are not\nvalid.\n\nExamples: input file is not of the expected type, is malformed\/corrupt, or\nwas flagged as malware"
          ],
          "type": "string",
          "enum": [
            "TEST_STATE_UNSPECIFIED",
            "VALIDATING",
            "PENDING",
            "RUNNING",
            "FINISHED",
            "ERROR",
            "UNSUPPORTED_ENVIRONMENT",
            "INCOMPATIBLE_ENVIRONMENT",
            "INCOMPATIBLE_ARCHITECTURE",
            "CANCELLED",
            "INVALID"
          ]
```

- [MatrixState.kt](https://github.com/Flank/flank/blob/master/test_runner/src/main/kotlin/ftl/util/MatrixState.kt)
- [testing_v1.json#L171](https://github.com/Flank/gcloud_cli/blob/master/google-cloud-sdk/lib/googlecloudsdk/third_party/apis/testing_v1.json#L171)

I think only:
```json
"VALIDATING",
"PENDING",
"RUNNING",
"FINISHED",
```
will may have a webLink that we can poll for.

```json
"TEST_STATE_UNSPECIFIED",
"UNSUPPORTED_ENVIRONMENT",
"INCOMPATIBLE_ENVIRONMENT",
"INCOMPATIBLE_ARCHITECTURE",
"CANCELLED",
"INVALID"
```

I don't think we'll get a web link from these states. In the current implementation, will we poll indefinitely until a web link is provided?

Reply:
Thanks for input. I will implement additional logic to prevent this.